lib/model/unreads.dart

```diff
@@ -0,0 +1,383 @@
+                // Happens as a bug in some cases before fixed in Zulip 6.0:
+                //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/unreads.20in.20unsubscribed.20streams/near/1458467
+                // TODO(server-6) remove Zulip 6.0 comment
+                return;
+              }
+              switch (detail.type) {
+                case MessageType.stream:
+                  ((newlyUnreadInStreams[detail.streamId!] ??= {})[detail.topic!] ??= QueueList())
+                    .add(messageId);
+                case MessageType.private:
+                  final narrow = DmNarrow.ofUpdateMessageFlagsMessageDetail(selfUserId: selfUserId,
+                    detail);
+                  (newlyUnreadInDms[narrow] ??= QueueList())
+                    .add(messageId);
+              }
+            }
+            newlyUnreadInStreams.forEach((incomingStreamId, incomingTopics) {
+              incomingTopics.forEach((incomingTopic, incomingMessageIds) {
+                _addAllInStreamTopic(incomingMessageIds..sort(), incomingStreamId, incomingTopic);
+              });
+            });
+            newlyUnreadInDms.forEach((incomingDmNarrow, incomingMessageIds) {
+              _addAllInDm(incomingMessageIds..sort(), incomingDmNarrow);
+            });
+        }
+    }
+    notifyListeners();
+  }
+
+  // TODO replace with efficient alternative
+  bool _slowIsPresentInStreams(int messageId) {
+    return streams.values.any(
+      (topics) => topics.values.any(
+        (messageIds) => messageIds.contains(messageId),
+      ),
+    );
+  }
+
+  void _addLastInStreamTopic(int messageId, int streamId, String topic) {
+    ((streams[streamId] ??= {})[topic] ??= QueueList()).addLast(messageId);
+  }
+
+  /// [messageIds] must be sorted ascending and without duplicates.
+  void _addAllInStreamTopic(QueueList<int> messageIds, int streamId, String topic) {
+    final topics = streams[streamId] ??= {};
+    QueueList<int> updatedMessageIds = messageIds;
+    if (topics[topic] != null) {
+      // setUnion dedupes existing and incoming unread IDs,
+      // so we tolerate zulip/zulip#22164, fixed in 6.0
+      // TODO(server-6) remove 6.0 comment
+      updatedMessageIds = setUnion(topics[topic]!, messageIds);
+    }
+    topics[topic] = updatedMessageIds;
+  }
+
+  // TODO replace with efficient alternative
+  void _slowRemoveAllInStreams(Iterable<int> messageIds) {
+    for (final messageId in messageIds) {
+      for (final stream in streams.values) {
+        for (final topic in stream.values) {
+          topic.remove(messageId);
+        }
+      }
+    }
+  }
+
+
+  void _removeInStreamTopic(int messageId, int streamId, String topic) {
+    streams[streamId]?[topic]?.remove(messageId);
+  }
+
+  // TODO (?) instead of making a one-off [Set], write and use `setDifference`
+  //   that operates on our kinds of sets (implemented as sorted lists)
+  void _removeAllInStreamTopic(Set<int> messageIds, int streamId, String topic) {
```

Comment:
It looks like we're only using this for DeleteMessageEvent. Those aren't common, I think, so no need to do anything fancy to optimize. (Especially as we're already, thankfully, doing this only for a single conversation's list of messages.)